### PR TITLE
Cleanup Annoy index

### PR DESCRIPTION
**Pull request by rschu1ze**
Some cleanup of the Annoy index. See the individual commits for details.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 4d0b7557f0d02d8e942ee9f662dff9c3b9cc67bd with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50312/4d0b7557f0d02d8e942ee9f662dff9c3b9cc67bd/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by rschu1ze** [comment]
@nickitat Not urgent but like to check this PR?

File: docs/en/engines/table-engines/mergetree-family/annindexes.md [hunk 1]
```diff
@@ -1,147 +1,156 @@
...
-  INDEX ann_index_name data TYPE ann_index_type(ann_index_parameters) GRANULARITY N
+  `embedding` Tuple(Float32[, Float32[, ...]]),
+  INDEX <ann_index_name> embedding TYPE <ann_index_type>(<ann_index_parameters>) GRANULARITY <N>
 )
 ENGINE = MergeTree
 ORDER BY id;
 ```
 
-With greater `GRANULARITY` indexes remember the data structure better. The `GRANULARITY` indicates how many granules will be used to construct the index. The more data is provided for the index, the more of it can be handled by one index and the more chances that with the right hyper parameters the index will remember the data structure better. But some indexes can't be built if they don't have enough data, so this granule will always participate in the query. For more information, see the description of indexes.
+ANN indexes are built during column insertion and merge and `INSERT` and `OPTIMIZE` statements will slower than for ordinary tables. ANNIndexes are ideally used only with immutable or rarely changed data, respectively comparatively many more read requests than write requests.
```
Comment (nickitat) [inline docs/en/engines/table-engines/mergetree-family/annindexes.md:57]:
```suggestion
ANN indexes are built during column insertion and merge and `INSERT` and `OPTIMIZE` statements will be slower than for ordinary tables. ANNIndexes are ideally used only with immutable or rarely changed data, respectively there are much more read requests than write requests.
```

File: src/Storages/MergeTree/ApproximateNearestNeighborIndexesCommon.cpp [hunk 2]
```diff
@@ -24,208 +22,166 @@ namespace ErrorCodes
...
-        return ANN::ANNQueryInformation::Metric::L2;
-    if (metric_name == "LpDistance")
-        return ANN::ANNQueryInformation::Metric::Lp;
-    return ANN::ANNQueryInformation::Metric::Unknown;
+    if (metric == "L2Distance")
+        return ApproximateNearestNeighborInformation::Metric::L2;
+    else if (metric == "LpDistance")
+        return ApproximateNearestNeighborInformation::Metric::Lp;
+    else
+        return ApproximateNearestNeighborInformation::Metric::Unknown;
```
Comment (nickitat) [inline src/Storages/MergeTree/ApproximateNearestNeighborIndexesCommon.cpp:49]: is it normal situation? shouldn't it also be an exception?

File: src/Storages/MergeTree/ApproximateNearestNeighborIndexesCommon.cpp [hunk 3]
```diff
@@ -24,208 +22,166 @@ namespace ErrorCodes
...
-                                 ContextPtr context) :
-    block_with_constants{KeyCondition::getBlockWithConstants(query_info.query, query_info.syntax_analyzer_result, context)},
-    ann_index_select_query_params{context->getSettings().get("ann_index_select_query_params").get<String>()},
-    index_granularity{context->getMergeTreeSettings().get("index_granularity").get<UInt64>()},
-    limit_restriction{context->getSettings().get("max_limit_for_ann_queries").get<UInt64>()},
-    index_is_useful{checkQueryStructure(query_info)} {}
+ApproximateNearestNeighborCondition::ApproximateNearestNeighborCondition(const SelectQueryInfo & query_info, ContextPtr context)
+    : block_with_constants(KeyCondition::getBlockWithConstants(query_info.query, query_info.syntax_analyzer_result, context))
+    , index_granularity(context->getMergeTreeSettings().get("index_granularity").get<UInt64>())
+    , max_limit_for_ann_queries(context->getSettings().get("max_limit_for_ann_queries").get<UInt64>())
```
Comment (nickitat) [inline src/Storages/MergeTree/ApproximateNearestNeighborIndexesCommon.cpp:57]: settings can be accessed by its names as data members.

File: src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp [hunk 4]
```diff
@@ -135,270 +138,254 @@ void MergeTreeIndexAggregatorAnnoy<Distance>::update(const Block & block, size_t
...
-        {
-            throw Exception(ErrorCodes::INCORRECT_DATA, "Can't parse first argument");
-        }
-    }
-    if (index.arguments.size() > 1 && !index.arguments[1].tryGet<String>(distance_name))
-    {
-        throw Exception(ErrorCodes::INCORRECT_DATA, "Can't parse second argument");
-    }
-    return std::make_shared<MergeTreeIndexAnnoy>(index, param, distance_name);
+    uint64_t trees = 100;
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:318]: maybe we could make it a parameter?

Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp:317]: Sure, done.

**Review by nickitat — APPROVED** [review]

**Pull request merged**